### PR TITLE
no op metadata json requests based on slot

### PR DESCRIPTION
**Pull request by Nagaprasadvr**
- [ ] add checks based slots while fetching metedata json ( restrict is data is not stale)
- [ ] add fetch_duration , fetch_failed_attempts, last_fetched_status_code to track metadata_json requests

### CODE REVIEW

File: core/src/metadata_json.rs [hunk 1]
```diff
@@ -90,9 +111,10 @@ impl MetadataJsonDownloadWorkerArgs {
                 }
 
                 let pool = pool.clone();
-                let client = client.clone();
 
-                handlers.push(spawn_task(client, pool, download_metadata_info));
+                if is_asset_data_fetch_req(&download_metadata_info, pool.clone()).await {
+                    handlers.push(spawn_task(client.clone(), pool, download_metadata_info));
+                }
```
Comment (kespinola) [inline core/src/metadata_json.rs:117]:
```suggestion
                if skip_index(&download_metadata_info, pool.clone()).await {
                  return
                }
                handlers.push(spawn_task(client.clone(), pool, download_metadata_info));
```

File: core/src/metadata_json.rs [hunk 2]
```diff
@@ -65,6 +67,25 @@ pub struct MetadataJsonDownloadWorkerArgs {
...
+        ..
+    } = download_metadata_info;
+
+    let conn = SqlxPostgresConnector::from_sqlx_postgres_pool(pool);
+
+    asset_data::Entity::find_by_id(asset_data_id.clone())
+        .one(&conn)
+        .await
+        .unwrap_or(None)
+        .is_some_and(|model| *incoming_slot > model.slot_updated)
```
Comment (kespinola) [inline core/src/metadata_json.rs:86]: The check is that the incoming url and slot < the incoming.

The other item to audit is the usage of reindex. These should be aligned so that this is done before they are marked for reindex and then just go off of it. 

Can it get in a spot where it never reaches the metaplex json downloader unless it needs to be downloaded?

File: core/src/metadata_json.rs [hunk 3]
```diff
@@ -188,7 +229,25 @@ async fn fetch_metadata_json(
...
     .retry(&ExponentialBuilder::default())
-    .await
+    .notify(|_e, _d| {
+        retries.fetch_add(1, Ordering::Relaxed);
+    })
+    .await;
+
+    let time_elapsed = start.elapsed().as_secs();
+
+    let retries = retries.load(Ordering::Relaxed);
```
Comment (kespinola) [inline core/src/metadata_json.rs:239]: backon give you access to the number of attempts?

File: core/src/metadata_json.rs [hunk 4]
```diff
@@ -206,22 +265,37 @@ pub async fn perform_metadata_json_task(
...
+    let conn = SqlxPostgresConnector::from_sqlx_postgres_pool(pool);
     match fetch_metadata_json(client, &download_metadata_info.uri).await {
         Ok(metadata) => {
             let active_model = asset_data::ActiveModel {
                 id: Set(download_metadata_info.asset_data_id.clone()),
-                metadata: Set(metadata),
+                metadata: Set(metadata.value),
                 reindex: Set(Some(false)),
+                last_requested_status_code: Set(Some(MetadataJsonFetchResult::Success)),
+                fetch_duration_in_secs: Set(Some(metadata.time_elapsed)),
```
Comment (kespinola) [inline core/src/metadata_json.rs:276]: Lets make this miliseconds

File: core/src/metadata_json.rs (hunk 3 above)
Comment (kespinola) [reply core/src/metadata_json.rs:239]: The function is a bit hard to follow with the match then decorating with extra context. Maybe do timing in the caller function and then have error respond with retries if can be retrieved from backon. If not then we can skip writing that to the db the status code is more informative. We can already tweak retries using backon so we know its max and timeout.

**Review by kespinola — CHANGES_REQUESTED** [review]
On a good track. We are missing migration for the new columns.

File: core/src/metadata_json.rs (hunk 2 above)
Comment (Nagaprasadvr) [reply core/src/metadata_json.rs:86]: this can be done using reindex flag , we only call download_metadata_notifier if reindex flag is set  . like here ?                                         
image

File: core/src/metadata_json.rs (hunk 3 above)
Comment (Nagaprasadvr) [reply core/src/metadata_json.rs]: backon doesnt give access to these , so dropping this from db

File: program_transformers/src/mpl_core_program/mod.rs [hunk 5]
```diff
@@ -30,6 +30,9 @@ pub async fn handle_mpl_core_account<'a, 'b, 'c>(
             )
             .await?
             {
+                if skip_metadata_json_download(&info, db).await {
+                    return Ok(());
+                }
                 download_metadata_notifier(info)
                     .await
                     .map_err(ProgramTransformerError::DownloadMetadataNotify)?;
```
Comment (kespinola) [inline program_transformers/src/mpl_core_program/mod.rs:38]: grpc-ingest and plerkle-ingest use different notifiers. If we can I'd like to apply this to grpc-ingest and ops. Idea here is to reduce the changes applied to plerkle with this update.

File: program_transformers/src/lib.rs [hunk 6]
```diff
@@ -254,3 +255,27 @@ fn record_metric(metric_name: &str, success: bool, retries: u32) {
...
+    asset_data::Entity::find_by_id(asset_data_id.clone())
+        .one(conn)
+        .await
+        .unwrap_or(None)
+        .is_some_and(|model| {
+            !model.reindex.unwrap_or(false)
+                && model.metadata_url.eq(uri)
+                && model.slot_updated >= *incoming_slot
+        })
+}
```
Comment (kespinola) [inline program_transformers/src/lib.rs:281]: I think the das-core crate metdata_json.rs is the spot for this function. Also recommend keeping it private if we can and just expose download_notifier.

File: digital_asset_types/src/dao/generated/sea_orm_active_enums.rs [hunk 7]
```diff
@@ -163,3 +163,16 @@ pub enum Instruction {
...
+    rs_type = "String",
+    db_type = "Enum",
+    enum_name = "last_requested_status_code"
+)]
+pub enum LastRequestedStatusCode {
+    #[sea_orm(string_value = "success")]
+    Success,
+    #[sea_orm(string_value = "failure")]
+    Failure,
+}
```
Comment (kespinola) [inline digital_asset_types/src/dao/generated/sea_orm_active_enums.rs:178]: I was thinking the different http status codes like 403 500, 200. Just saving the status code number instead will suffice.

**Review by kespinola — CHANGES_REQUESTED** [review]
This is going to be annoying but I would like to add the updates to the db in a PR after grpc-ingest is merged in. As I noted in another comment we want to limit the impact of this PR on plerkle-ingest which means trying to limit our updates to program_transforms or the database model.
